File: 02_binary-search-tree.rb
###############################################################################
#
# 2. Write ruby code that models the following (Binary Search Tree):
# 
#     left child is always less than the parent & right child is always bigger than
#     the parent.
# 
#     A parent can only have 2 immediate child node.
# 
#     1. Write a method to add item to the model so caller can build the tree
#        like below. [see Binary_Search_Tree instance method: add]
#     2. Then perform a walk of depth first search on the model so we get result
#        like this: `21, 56, 62, 67, 78, 81, 97, 115`
#        [see Binary_Search_Tree instance method: pre_order]
# 
#     ![Binary Search Tree example](images/Binary_Search_Tree_example.jpg)
#
################################################################################

class Node
  def initialize(value, left = nil, right = nil)
    @value = value
    @left = left
    @right = right
  end

  def value
    @value
  end

  def value=(value)
    @value = value
  end

  def left
    @left
  end

  def left=(node)
    @left = node
  end

  def right
    @right
  end

  def right=(node)
    @right = node
  end
end

class Binary_Search_Tree
  def initialize
    @node = nil
  end

  def add(new_value)
    # If there's no node here yet, create a new one for the new value.
    node = @node
    if (node == nil)
      @node = Node.new(new_value)
      return nil
    end

    # If we get here, there was already a node at this level, so search for the
    # right spot to insert a node with the new value.
    return insert_value(node, new_value)
  end

  def pre_order
    result = []
    traverse_pre_order(@node, result)
    result
  end

  ## Helpers

  # Find an empty spot in the tree and insert a new node with the specified
  # value.
  def insert_value(node, value)
    # If the provided value is lower than the current node's value, search the
    # left side for an empty spot.
    if value < node.value
      # If there's no value to the left of the current node, insert the
      # provided value there.
      if (node.left == nil)
        node.left = Node.new(value)
        return nil
      end

      # Otherwise, dig deeper
      return self.insert_value(node.left, value)

    # If the provided value is higher than the value of the current node, look
    # for an empty spot on the right side of the node.
    elsif value > node.value
      # If there's no value to the right of the current node, insert the
      # provided value there.
      if (node.right == nil)
        node.right = Node.new(value)
        return nil
      end

      # Otherwise, dig deeper
      return insert_value(node.right, value)
    end
  end

  def traverse_pre_order(node, result)
    result << node.value
    self.traverse_pre_order(node.left, result) if node.left != nil
    self.traverse_pre_order(node.right, result) if node.right != nil
  end
end

values = [21, 56, 62, 67, 78, 81, 97, 115]

bst = Binary_Search_Tree.new

values.each { |value| bst.add(value) }

p bst.pre_order

File: 02_binary-search-tree.js
/* eslint-disable */
/*******************************************************************************
 * 
 * 2. Write ruby code that models the following (Binary Search Tree):
 * 
 *    left child is always less than the parent & right child is always bigger
 *    than the parent.
 * 
 *    A parent can only have 2 immediate child nodes.
 * 
 *    1. Write a method to add an item to the model so the caller can build the
 *       tree like below. BST. [see BST instance method: add]
 *    2. Then perform a walk of depth first search on the model so we get result
 *       like this: `21, 56, 62, 67, 78, 81, 97, 115`
 *       [see BST instance method: preOrder]
 * 
 *    ![Binary Search Tree example](images/Binary_Search_Tree_example.jpg)
 * 
 *******************************************************************************
*/
/* eslint-enable */


// One node of the Binary Search Tree
class Node {
  constructor(value, left = null, right = null) {
    this.value = value;
    this.left = left;
    this.right = right;
  }
}


// BST is the Binary Search Tree that contains all of the nodes.
class BST {
  constructor() {
    this.node = null;
  }

  add(newValue) {
    // If there's no node here yet, create a new one for the new value.
    const node = this.node;
    if (node === null) {
      this.node = new Node(newValue);
      return;
    }

    // There's a node at this level, so search for the right spot to create a
    // new node for the new value.
    const insertValue = function(node, newValue) {
      // If the new value is lower than the value at the current level, search
      // the left side for an empty place.
      if (newValue < node.value) {
        // If there's no value to the left of the current node, insert the
        // provided value there.
        if (node.left === null) {
          node.left = new Node(newValue);
          return;
        }

        // Otherwise, dig deeper
        return insertValue(node.left, newValue);

      // If the new value is higher than the value of the current node, look for
      // an empty spot on the right.
      } else if (newValue > node.value) {
        if (node.right === null) {
          node.right = new Node(newValue);
          return;
        }

        // Otherwise, dig deeper
        return insertValue(node.right, newValue);
      }
    };

    return insertValue(node, newValue);
  }

  preOrder() {
    if (this.node === null) {
      return null;
    } else {
      const result = [];
      const traversePreOrder = (node) => {
        result.push(node.value);
        node.left && traversePreOrder(node.left);
        node.right && traversePreOrder(node.right);
      };
      traversePreOrder(this.node);
      return result;
    };
  }
}

const bst = new BST();
const values = [21, 56, 62, 67, 78, 81, 97, 115];

values.forEach((value) => bst.add(value));

console.log(bst.preOrder());

